### PR TITLE
TOOL-23653 Upgrade version of Python on DCoL

### PR DESCRIPTION
**Pull request by sanjipta-delphix**
<details open>
<summary><h2> Problem </h2></summary>

We are planning to upgrade python in dcol machines. This will unblock us for upgrading ansible. Since we are using ubuntu 20.04, we should add deadsnakes PPA.
</details>

<details open>
<summary><h2> Solution </h2></summary>

I have tested new linux package mirror after adding deadsnake PPA.
PR: https://github.com/delphix/devops-gate/pull/2572.

Python 3.12 version is now available. So in this PR I have updated the python version to 3.12.

</details>

<details>
<summary><h2> Testing Done </h2></summary>

I have tested the new python version in a test VM in DCoA ubuntu-linux-mirror-package-sanjipta .
</details>

### CODE REVIEW

File: live-build/variants/internal-dcenter/package-lists/dcenter.list.chroot [hunk 1]
```diff
@@ -26,7 +26,7 @@ nfs-common-dbgsym
 nfs-kernel-server
 nfs-kernel-server-dbgsym
 openjdk-11-jdk-headless
-python3
+python3.12
 python3-dbg
 python3-dev
 python3-venv
```
Comment (grodr) [inline live-build/variants/internal-dcenter/package-lists/dcenter.list.chroot]: I think these three packages also need the python version bumped so that we are installing the python3.12 compatible versions of the packages.

**Review by grodr — COMMENTED** [review]
Looking good! To test this I think we'll need to run the appliance-build Jenkins job with these changes and then validate that we can run the dcol package on the dcenter variant produced by that build. It looks like the job kicked off by your PR here failed https://github.com/delphix/appliance-build/pull/749/checks?check_run_id=20900968283 I am guessing that is because the devops-gate change to update the package mirror has not been integrated yet. I think we can integrate that change and rerun your check job.

File: live-build/variants/internal-dcenter/package-lists/dcenter.list.chroot (hunk 1 above)
Comment (ghost) [inline live-build/variants/internal-dcenter/package-lists/dcenter.list.chroot]: I think we'll want to use the 3.12 version of these packages as well

File: live-build/variants/internal-dcenter/package-lists/dcenter.list.chroot
Comment (ghost) [inline live-build/variants/internal-dcenter/package-lists/dcenter.list.chroot:1]: For further testing can we generate a DCoL from this, or build an image and apply it to dcol-dev?

**Review by ghost — COMMENTED** [review]
Looks great1

File: live-build/variants/internal-dcenter/package-lists/dcenter.list.chroot (hunk 1 above)
Comment (sanjipta-delphix) [reply live-build/variants/internal-dcenter/package-lists/dcenter.list.chroot]: I have updated with python3.12

Comment (sanjipta-delphix) [reply live-build/variants/internal-dcenter/package-lists/dcenter.list.chroot]: I have updated with python3.12.

File: .github/workflows/main.yml
Comment (ghost) [inline .github/workflows/main.yml:1]: This change seems unrelated to the task at hand, could it be broken out into a separate PR?

**Review by palash-gandhi — CHANGES_REQUESTED** [review]
The upstream branch of this PR is set to `master` which is now defunct. You should change this to `develop` so that your `appliance-build` PR check can pass.

File: .github/workflows/main.yml
Comment (sanjipta-delphix) [reply .github/workflows/main.yml:1]: Hi @jtamagnan-delphix ,
When it was ubuntu-18.04, it wasn't getting runners and throwing error-  " This request was automatically failed because there were no enabled runners online to process the request for more than 1 days".

So I updated to ubuntu-20.04 version.

**Comment by grodr** [comment]
For testing is there a plan to deploy this to dcol-dev to validate that the dcenter code and basic operations work properly on the new image?

**Comment by ghost** [comment]
> For testing is there a plan to deploy this to dcol-dev to validate that the dcenter code and basic operations work properly on the new image?

Yes, tomorrow Sanjipta and I will be deploying this to dcol-dev to test that it works as expected

File: live-build/variants/internal-dcenter/package-lists/dcenter.list.chroot [hunk 2]
```diff
@@ -24,8 +24,10 @@ libldap2-dev
 libsasl2-dev
 nfs-kernel-server
 openjdk-11-jdk-headless
-python3
-python3-dev
+python3.12
+python3.12-dbg
+python3.12-dev
 python3-venv
+python3.12-venv
```
Comment (ghost) [inline live-build/variants/internal-dcenter/package-lists/dcenter.list.chroot:30]: This line looks duplicated

File: live-build/variants/internal-dcenter/package-lists/dcenter.list.chroot [hunk 3]
```diff
@@ -24,8 +24,10 @@ libldap2-dev
 libsasl2-dev
 nfs-kernel-server
 openjdk-11-jdk-headless
-python3
-python3-dev
+python3.12
+python3.12-dbg
```
Comment (ghost) [inline live-build/variants/internal-dcenter/package-lists/dcenter.list.chroot]: We aren't using the `python3-dbg` package. i imagine we might not need the python3.12 version of it either

File: live-build/variants/internal-dcenter/package-lists/dcenter.list.chroot (hunk 2 above)
Comment (sanjipta-delphix) [reply live-build/variants/internal-dcenter/package-lists/dcenter.list.chroot:30]: @jtamagnan-delphix ,
You asked me earlier to keep both python3-venv and python3.12-venv?

Comment (ghost) [reply live-build/variants/internal-dcenter/package-lists/dcenter.list.chroot:30]: Apologies I thought I had read `python3.12-venv` twice

File: live-build/variants/internal-dcenter/package-lists/dcenter.list.chroot (hunk 3 above)
Comment (sanjipta-delphix) [reply live-build/variants/internal-dcenter/package-lists/dcenter.list.chroot]: @jtamagnan-delphix , removed this.

File: live-build/misc/ansible-roles/appliance-build.dcenter/tasks/main.yml [hunk 4]
```diff
@@ -27,6 +27,16 @@
...
 
+- alternatives:
+    name: python3
+    link: /usr/bin/python3
+    path: /usr/bin/python3.12
+
+- alternatives:
+    name: python3
+    link: /bin/python3
+    path: /bin/python3.12
```
Comment (ghost) [inline live-build/misc/ansible-roles/appliance-build.dcenter/tasks/main.yml]: I think we can remove this one

**Review by ghost — APPROVED** [review]

File: live-build/misc/ansible-roles/appliance-build.dcenter/tasks/main.yml (hunk 4 above)
Comment (sanjipta-delphix) [reply live-build/misc/ansible-roles/appliance-build.dcenter/tasks/main.yml]: @jtamagnan-delphix , I have removed this.

**Review by palash-gandhi — APPROVED** [review]
